File: src/solid/dependency-inversion.ts
enum Relationship {
  parent,
  child,
  sibling,
}

interface IRelationship {
  from: Person;
  type: Relationship;
  to: Person;
}

class Person {
  constructor(public name: string) {}
}

// to fixt the dependency relationship between low and high level
// we use an abstract class
abstract class RelationshipBrowser {
  abstract findAllChildrenOf(name: string): Person[];
}

// LOW-LEVEL MODULE
class Relationships extends RelationshipBrowser {
  public data: IRelationship[];
  constructor() {
    super();
    this.data = [];
  }

  addParentAndChild(parent: Person, child: Person) {
    this.data.push({
      from: parent,
      type: Relationship.parent,
      to: child,
    });
  }
  // this api should always gives the right result even if the data structure changes
  findAllChildrenOf(name: string) {
    return this.data
      .filter((r) => r.from.name === name && r.type === Relationship.parent)
      .map((r) => r.to);
  }
}

// HIGH-LEVEL MODULE
class Research {
  // dependecy problem
  // constructor(relationships: Relationships) {
  //   // find all children of John
  //   let relations = relationships.data;
  //   for (let rel of relations.filter(
  //     (r) => r.from.name === 'John' && r.type === Relationship.parent
  //   )) {
  //     console.log(`John has a child named ${rel.to.name}`);
  //   }

  // dependecy abstraction implementention
  constructor(browser: RelationshipBrowser) {
    for (let p of browser.findAllChildrenOf('John')) {
      console.log(`John has a child called ${p.name}`);
    }
  }
}

let parent1 = new Person('John');
let child1 = new Person('Chris');
let child2 = new Person('Matt');

let rels = new Relationships();
rels.addParentAndChild(parent1, child1);
rels.addParentAndChild(parent1, child2);

new Research(rels);

File: python/solid/dependency-inversion.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Any
from enum import Enum


class Relationship(Enum):
    parent = 'parent'
    child = 'child'
    sibling = 'sibling'

class Relation:
    def __init__(self, from_person: Person, to_person: Person, relationship: Relationship):
        self.from_person = from_person
        self.to_person = to_person
        self.relationship = relationship

class Person:
    def __init__(self, name: str):
        self.name = name

class RelationshipBrowser(ABC):
    @abstractmethod
    def find_all_children_of(self, name: str) -> list[Person]:
        pass

class Relationships(RelationshipBrowser):
    def __init__(self):
        self.relations = []
    
    def add_parent_and_child(self, parent: Person, child: Person):
        self.relations.append(Relation(parent, child, Relationship.parent))
        self.relations.append(Relation(child, parent, Relationship.child))
    
    def find_all_children_of(self, name: str) -> list[Person]:
        person_objects = filter(lambda r: r.from_person.name == name and r.relationship == Relationship.parent, self.relations)
        return list(map(lambda r: r.to_person, person_objects))

class Research:
    def __init__(self, browser: RelationshipBrowser):
        for p in browser.find_all_children_of('John'):
            print(f'John has a child called {p.name}')

parent = Person('John')
child1 = Person('Chris')
child2 = Person('Matt')

relationships = Relationships()
relationships.add_parent_and_child(parent, child1)
relationships.add_parent_and_child(parent, child2)

Research(relationships)